vm-virtio/src/vhost_user/net.rs

```diff
@@ -0,0 +1,315 @@
+
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateKillEventFd)?;
+
+        let mut avail_features = 1 << virtio_net::VIRTIO_NET_F_GUEST_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_MRG_RXBUF
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_EVENT_IDX;
+
+        // Get features from backend, do negotiation to get a feature collection which
+        // both VMM and backend can support.
+        let backend_features = vhost_user_net.get_features().unwrap();
+        avail_features &= backend_features;
+        // Set features back here is decided by the vhost crate mechanism, since the 
+        // later vhost call requires backend_features filled in master as a pre-requirement,
+        // which is setup by the call here. Will check if the corresponding logic in vhost
+        // is sensible in the future.
+        vhost_user_net
+            .set_features(backend_features)
+            .map_err(Error::VhostUserSetFeatures)?;
+
+        let mut config_space = Vec::with_capacity(MAC_ADDR_LEN);
+        unsafe { config_space.set_len(MAC_ADDR_LEN) }
+        config_space[..].copy_from_slice(mac_addr.get_bytes());
+        avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC;
+
+        let mut acked_features = 0;
+        if backend_features & (1 << VHOST_USER_F_PROTOCOL_FEATURES) != 0 {
+            acked_features |= 1 << VHOST_USER_F_PROTOCOL_FEATURES;
+            let protocol_features = vhost_user_net.get_protocol_features().unwrap();
+            vhost_user_net
+                .set_protocol_features(protocol_features)
+                .map_err(Error::VhostUserSetProtocolFeatures)?;
+        }
+
+        vhost_user_net
+            .set_owner()
+            .map_err(Error::VhostUserSetOwner)?;
+
+        // Send set_vring_base here, since it could tell backends, like OVS + DPDK,
+        // how many virt queues to be handled, which backend required to know at early stage.
+        for i in 0..num_queues {
+            vhost_user_net
+                .set_vring_base(i, 0)
+                .map_err(Error::VhostUserSetVringBase)?;
+        }
+
+        Ok(Net {
+            vhost_user_net,
+            kill_evt,
+            avail_features,
+            acked_features,
+            backend_features,
+            config_space,
+            queue_sizes: vec![queue_size; num_queues],
+        })
+    }
+
+    pub fn setup_vunet(
```

Comment:
Same here, this function should not be specific to vhost-user-net. Unless I'm missing something, it should be factorized at the module level so that all submodules can benefit from it.
This means `fs.rs` needs to be moved from `vm-virtio/src/` to `vm-virtio/src/vhost_user/` as a prerequisite.